File: Day-01-Arrays/CPP/53_Maximum_Subarray.cpp
/*
===============================================================================
                            DSA Interview Playbook
===============================================================================

Problem:
53. Maximum Subarray

Link:
https://leetcode.com/problems/maximum-subarray/

Pattern:
Kadane's Algorithm

Difficulty:
Medium

Time Complexity:
O(n)

Space Complexity:
O(1)

===============================================================================
*/

class Solution {
public:
    int maxSubArray(vector<int>& nums) {

        int currentSum = nums[0];
        int maxSum = nums[0];

        for (int i = 1; i < nums.size(); i++) {
            currentSum = max(nums[i], currentSum + nums[i]);
            maxSum = max(maxSum, currentSum);
        }

        return maxSum;
    }
};

File: 01-Arrays/Python/53_Maximum_Subarray.py
"""
Problem: 53. Maximum Subarray

Pattern: Kadane's Algorithm

Time Complexity: O(n)
Space Complexity: O(1)
"""

class Solution:
    def maxSubArray(self, nums: list[int]) -> int:

        currentSum = nums[0]
        maxSum = nums[0]

        for num in nums[1:]:
            currentSum = max(num, currentSum + num)
            maxSum = max(maxSum, currentSum)

        return maxSum